src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -137,94 +144,177 @@ boolean removeUnreachableBlocks ()
 				}
 			}
 		}
-
 		final Set<L2BasicBlock> unreachableBlocks = new HashSet<>(blocks);
 		unreachableBlocks.removeAll(reachableBlocks);
 		for (final L2BasicBlock block : unreachableBlocks)
 		{
 			block.instructions().forEach(L2Instruction::justRemoved);
 			block.instructions().clear();
 		}
-		return blocks.retainAll(reachableBlocks);
+		final boolean changed = blocks.retainAll(reachableBlocks);
+		// See if any blocks no longer need to be a loop head.
+		if (changed)
+		{
+			for (final L2BasicBlock block : blocks)
+			{
+				if (block.isLoopHead
+					&& block.predecessorEdgesCopy().stream()
+						.noneMatch(L2PcOperand::isBackward))
+				{
+					// It's a loop head that has no back-edges pointing to it.
+					block.isLoopHead = false;
+				}
+			}
+		}
+		return changed;
 	}
 
 	/**
-	 * Given the set of instructions which are reachable, compute the needed
-	 * subset, which consists of those which have side-effect or produce a value
-	 * consumed by other needed instructions.  Don't assume SSA.
+	 * Remove any unnecessary instructions.  Answer true if any were removed.
 	 *
-	 * @return The instructions that are needed and should be kept.
+	 * @param dataCouplingMode
+	 *        The {@link DataCouplingMode} that chooses how to trace liveness.
+	 * @return Whether any dead instructions were removed or changed.
 	 */
-	private Set<L2Instruction> findNeededInstructions ()
+	private boolean removeDeadInstructions (
+		final DataCouplingMode dataCouplingMode)
 	{
-		final Deque<L2Instruction> instructionsToVisit = new ArrayDeque<>();
+		final DeadCodeAnalyzer analyzer =
+			new DeadCodeAnalyzer(dataCouplingMode, controlFlowGraph);
+		analyzer.analyzeReads();
+		final Set<L2Instruction> liveInstructions = analyzer.liveInstructions();
```

Comment:
Ah, yes, here's the code that I was predicting had to be somewhere.  Looks like the old code iterated in reverse order, and the new code iterates in forward order. I was expecting the former behavior, but it's encouraging that you can use the latter, as it suggests greater cleanliness of the CFG at this point. Maybe this is necessary to support your conditional replacement mechanism.

Reply:
I believe it's order-independent, since afterward it visits each manifest, discarding registers from the definitions map's values (which are lists of registers) which no longer have any definitions (writes).  There's another suggestion in this change set, for which I've updated a comment about this.